### ISSUE TITLE
Provide lto plugin in bfd-plugins to make ar able to handle files compiled with -flto

### ISSUE THREAD

**Issue by db4ple** [gnu-mcu-eclipse/arm-none-eabi-gcc#4]
Well, subject says it all.
If you compile a file with
```
arm-none-eabi-gcc.exe -flto -o dummy.o dummy.c
ar -r libdummy.a dummy.o
```
you get the following error:
```
dummy.o: plugin needed to handle lto object
```

In the ARM GCC builds there is inside lib/gcc/arm-none-eabi/<version> a liblto_plugin-0.dll next to lto1.exe which one can copy into bfd-plugins (lib/bfd-plugins), which makes the error message go away. 

It would be very nice and helpful to have the same out of the box working with this build as I really like the xpm based "no settings required" approach.

**Comment by ilg-ul** [comment]
You mean in the ARM distribution this file is present? Then it should be posible to add it to the GME pacakge too.

I'll investigate.

**Comment by ilg-ul** [comment]
I don't have a very good experience with lto, each time I tried it, something failed. It seems you managed to make it run. Could you share your experience?

**Comment by db4ple** [comment]
ARM GCC:
Yes, it is present there and it does work. I got the idea from https://bugs.launchpad.net/gcc-arm-embedded/+bug/1808722 , btw.

Using lto in general:
In the UHSDR project (https://github.com/df8oe/UHSDR) we use -flto extensively, all our builds (make and Eclipse) are using it without problems in the field. 

In Eclipse we have to build a library in order to get around the Windows issue of command line length limitation. Here we build at the moment only the main application with -flto and a library without lto and this mix also works without problems. But of course it would be nicer to build everything with -flto.

**Comment by ilg-ul** [comment]
> create the bfd-plugin dir manually and copy the dll over from the other directory.

what if we create this folder when the distribution is packed?

**Comment by ilg-ul** [comment]
> in order to get around the Windows issue of command line length 

btw, in the xPack builder I plan to use the @file feature of gcc, i.e. if the command line is longer than 8K, to store the command options in a file and instruct gcc to use it.

it is not yet implemented, but it is on my TODO list.

**Comment by db4ple** [comment]
> 
> 
> > create the bfd-plugin dir manually and copy the dll over from the other directory.
> 
> what if we create this folder when the distribution is packed?
I think this should work. But as far as I can see, the main issue is that a dynamically linkable lto plugin must be produced by the build process. Or, and that is not my area of expertise (I build my last gcc and binutils from source almost two decades ago...) the binutils are build with the plugin statically linked in.
Whatever works better/is easier, I don't care...

**Comment by ilg-ul** [comment]
> the main issue is that a dynamically linkable lto plugin must be produced by the build process.

yes, the build scripts are very complicated, but if ARM did it, I'll take a look at their scripts and try to find a solution. there is no guarantee yet that this is resonable, but I'll take a look these days.

> lib/gcc/arm-none-eabi/liblto_plugin-0.dll

Is the file in this location used? Perhaps the ARM script misplaced it, and it should be located only in the `bfd-plugin` folder.

**Comment by ilg-ul** [comment]
I did a quick LTO test on macOS with my project templates, and the build is ok, but the resulting executable crashes QEMU from the very beginning. I have to investigate if the problem is in the executable or in QEMU, but I would not rule out a problem in the executable. 

Did you have to do anything special in the startup to support LTO?

**Comment by db4ple** [comment]
Not really anything special we are aware of, but we are using our own startup code (based on the STM32F4 template).  It would have to be investigated where/when exactly your code crashes in QEMU. 

I just recently tried our code in QEMU and it does start. Well it doesn't run well, because we're not passing the PLL startup procedure (we get stuck in the error handler in an endless loop) but that is after we enter main() and most likely a problem of the emulation as such.  Not a crash.

**Comment by ilg-ul** [comment]
> It would have to be investigated where/when exactly your code crashes in QEMU

will do.

the usual suspects for LTO are code reordering and code removal. inspecting the program listing can help. however this does not justify the QEMU crash.

**Comment by ilg-ul** [comment]
I checked and the problem occurs in the new version of GDB. previous version is ok.

https://bugs.launchpad.net/gcc-arm-embedded/+bug/1813553

So, if you upgrade to GCC 8.2, you need to still use GDB from 7.x.

I also took a look at the generated code. It is massively reordered, functions called only once are inlined, etc, but I did not identify any problems, as long as the reset vectors are in the correct place, the the application should work.

Could you confirm using GCC 7.x?

---

Next to check is the liblto_plugin problem.

**Comment by ilg-ul** [comment]
I fixed the build script to create the soft link on POSIX systems, `ar` now can create LTO archives:

```
make all 
Building file: ../file.c
Invoking: GNU ARM Cross C Compiler
arm-none-eabi-gcc -mcpu=cortex-m3 -mthumb -O0 -fmessage-length=0 -fsigned-char -ffunction-sections -fdata-sections -flto  -g3 -std=gnu11 -MMD -MP -MF"file.d" -MT"file.o" -c -o "file.o" "../file.c"
Finished building: ../file.c
 
Building target: libstatic-lto.a
Invoking: GNU ARM Cross Archiver
arm-none-eabi-ar -r  "libstatic-lto.a"  ./file.o   
arm-none-eabi-ar: creating libstatic-lto.a
Finished building target: libstatic-lto.a
```

---

Next I'll see what can be done for Windows.

**Comment by db4ple** [comment]
Cool!  I  think on Windows you may have to copy the plugin.

We are using both gcc 8.2.1 and 7.x.x with LTO with success. We in general don't use QEMU but of course GDB with Eclipse. The new GCC 8.2.x does optimize some parts a whole lot better, shrinks our binary on the STM32F4 by 70k (interestingly almost no change for F7 and H7, but that is a different story).

**Comment by ilg-ul** [comment]
> in general don't use QEMU

so far I could not detect any problem with QEMU, the crash was only related to GDB 8.2, after replacing it with an older GDB, my blinky application ran just fine in QEMU.

> We are using both gcc 8.2.1 and 7.x.x with LTO with success

it looks like you found a winning combinations of options if you can debug with the GDB packed by 8.2.

could you check and share the details?

**Comment by db4ple** [comment]
I did not use the gdb from the xpacks yet. And thinking of it, no one on Windows was probably using the GDB in 8.2.1 for the simple reason that the GCC 8.2.1 from ARM GCC bails out with an error at some point during the linking. But on Linux some use the GCC 8.2.1 but I don't know if anyone used the GDB so it might be the case that the GDB 8.2.1 is really broken. I'll try that but it might take a few days.

**Comment by ilg-ul** [comment]
> I think on Windows you may have to copy the plugin

unfortunately it isn't that simple. I ran a build and the .dll, for one reason or another, is not built at all. :-(

I have to find what is missing in the mingw build, that prevents this file to be created. this might take some time...

the linux/macOS builds properly create the plugin.

**Comment by db4ple** [comment]
Hi,
yes, project is essentially C. Of course, with C++ it is a little more tricky, with global objects and their constructors/destructors and all the rest.

BTW, I do appreciate the work you put into my request!

**Comment by ilg-ul** [comment]
you were right, the missing dll on Windows was caused by the static build. I tweaked the build script and now the dll is created; I later copy it to the bfd-plugins folder for ar to find it. on posix I create soft links.

the windows binaries are *mostly* static, except libwinpthread-1.dll, which I copy to the bin folder.

if everything goes as expected, tomorrow evening I'll have the new binaries, run tests and on Sunday publish them.

**Comment by db4ple** [comment]
+1

**Comment by ilg-ul** [comment]
I have the new Linux and Windows binaries, AR is happy, GDB is happy, and generally Linux and macOS are ok.

the bad news is that I ran into a major problem on Windows, where my project fails to link entirely if I enable LTO, and passes if I disable it:

```
COMPILER_PATH=c:/users/ilg/appdata/roaming/gnu mcu eclipse/arm embedded gcc/8.2.1-1.3-20190202-1016/lib/gcc/../../libexec/gcc/arm-none-eabi/8.2.1/;c:/users/ilg/appdata/roaming/gnu mcu eclipse/arm embedded gcc/8.2.1-1.3-20190202-1016/lib/gcc/../../libexec/gcc/;c:/users/ilg/appdata/roaming/gnu mcu eclipse/arm embedded gcc/8.2.1-1.3-20190202-1016/bin/../libexec/gcc/arm-none-eabi/8.2.1/;c:/users/ilg/appdata/roaming/gnu mcu eclipse/arm embedded gcc/8.2.1-1.3-20190202-1016/bin/../libexec/gcc/arm-none-eabi/8.2.1/;c:/users/ilg/appdata/roaming/gnu mcu eclipse/arm embedded gcc/8.2.1-1.3-20190202-1016/bin/../libexec/gcc/;c:/users/ilg/appdata/roaming/gnu mcu eclipse/arm embedded gcc/8.2.1-1.3-20190202-1016/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/;c:/users/ilg/appdata/roaming/gnu mcu eclipse/arm embedded gcc/8.2.1-1.3-20190202-1016/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/
LIBRARY_PATH=c:/users/ilg/appdata/roaming/gnu mcu eclipse/arm embedded gcc/8.2.1-1.3-20190202-1016/bin/../lib/gcc/arm-none-eabi/8.2.1/thumb/v7e-m/nofp/;c:/users/ilg/appdata/roaming/gnu mcu eclipse/arm embedded gcc/8.2.1-1.3-20190202-1016/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/lib/thumb/v7e-m/nofp/;c:/users/ilg/appdata/roaming/gnu mcu eclipse/arm embedded gcc/8.2.1-1.3-20190202-1016/bin/../arm-none-eabi/lib/thumb/v7e-m/nofp/;c:/users/ilg/appdata/roaming/gnu mcu eclipse/arm embedded gcc/8.2.1-1.3-20190202-1016/bin/../lib/gcc/arm-none-eabi/8.2.1/;c:/users/ilg/appdata/roaming/gnu mcu eclipse/arm embedded gcc/8.2.1-1.3-20190202-1016/bin/../lib/gcc/;c:/users/ilg/appdata/roaming/gnu mcu eclipse/arm embedded gcc/8.2.1-1.3-20190202-1016/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/lib/;c:/users/ilg/appdata/roaming/gnu mcu eclipse/arm embedded gcc/8.2.1-1.3-20190202-1016/bin/../arm-none-eabi/lib/
COLLECT_GCC_OPTIONS='-c' '-fno-openmp' '-fno-openacc' '-mcpu=cortex-m4' '-mfloat-abi=soft'  '-Og' '-mcpu=cortex-m4' '-mthumb' '-mfloat-abi=soft' '-Og' '-fmessage-length=0' '-fsigned-char' '-ffunction-sections' '-fdata-sections' '-fno-move-loop-invariants' '-Wextra' '-g3' '-T' 'mem.ld' '-T' 'libs.ld' '-T' 'sections.ld' '-nostartfiles' '-L../ldscripts' '-specs=nano.specs' '-v'  '-dumpdir' './' '-dumpbase' 'f4b-lto.elf.wpa' '-fltrans-output-list=C:\Users\ilg\AppData\Local\Temp\ccE7EbYa.ltrans.out' '-fwpa' '-fresolution=C:\Users\ilg\AppData\Local\Temp\cclQK003.res' '-flinker-output=exec' '-march=armv7e-m'
arm-none-eabi-g++ -mcpu=cortex-m4 -mthumb -mfloat-abi=soft -march=armv7e-m -r -nostdlib -o C:\Users\ilg\AppData\Local\Temp\ccwLlI8qdebugobj C:\Users\ilg\AppData\Local\Temp\ccO00soGdebugobjtem C:\Users\ilg\AppData\Local\Temp\cciQzGiYdebugobjtem C:\Users\ilg\AppData\Local\Temp\cco9bNpidebugobjtem C:\Users\ilg\AppData\Local\Temp\ccYdpCQBdebugobjtem C:\Users\ilg\AppData\Local\Temp\ccSqTGUXdebugobjtem C:\Users\ilg\AppData\Local\Temp\ccWSaiqldebugobjtem C:\Users\ilg\AppData\Local\Temp\ccy2HMfLdebugobjtem C:\Users\ilg\AppData\Local\Temp\cc0q94Gddebugobjtem C:\Users\ilg\AppData\Local\Temp\cce4tsVJdebugobjtem C:\Users\ilg\AppData\Local\Temp\cci6FHridebugobjtem C:\Users\ilg\AppData\Local\Temp\ccsbQriRdebugobjtem C:\Users\ilg\AppData\Local\Temp\ccCCp15odebugobjtem C:\Users\ilg\AppData\Local\Temp\ccQZBICZdebugobjtem C:\Users\ilg\AppData\Local\Temp\cci6rOazdebugobjtem C:\Users\ilg\AppData\Local\Temp\cc6HcQqcdebugobjtem C:\Users\ilg\AppData\Local\Temp\cc2z9qHPdebugobjtem C:\Users\ilg\AppData\Local\Temp\cccvqBOvdebugobjtem C:\Users\ilg\AppData\Local\Temp\ccUznffedebugobjtem C:\Users\ilg\AppData\Local\Temp\ccojRpoYdebugobjtem C:\Users\ilg\AppData\Local\Temp\ccIvCmsLdebugobjtem C:\Users\ilg\AppData\Local\Temp\ccshTZUAdebugobjtem C:\Users\ilg\AppData\Local\Temp\ccAZhTDodebugobjtem C:\Users\ilg\AppData\Local\Temp\ccUpq4ggdebugobjtem C:\Users\ilg\AppData\Local\Temp\ccZLMx2udebugobjtem C:\Users\ilg\AppData\Local\Temp\ccomHN4Vdebugobjtem C:\Users\ilg\AppData\Local\Temp\ccV4Z6oqdebugobjtem C:\Users\ilg\AppData\Local\Temp\ccuSIQyVdebugobjtem
c:/users/ilg/appdata/roaming/gnu mcu eclipse/arm embedded gcc/8.2.1-1.3-20190202-1016/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld.exe: error: C:\Users\ilg\AppData\Local\Temp\ccO00soGdebugobjtem: ELF section name out of range
collect2.exe: error: ld returned 1 exit status
lto-wrapper.exe: fatal error: arm-none-eabi-g++ returned 1 exit status
compilation terminated.
c:/users/ilg/appdata/roaming/gnu mcu eclipse/arm embedded gcc/8.2.1-1.3-20190202-1016/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld.exe: error: lto-wrapper failed
collect2.exe: error: ld returned 1 exit status
make: *** [makefile:64: f4b-lto.elf] Error 1
"make all" terminated with exit code 2. Build might be incomplete.
```

I entered a new bug report:

https://bugs.launchpad.net/gcc-arm-embedded/+bug/1814397


I'm afraid LTO is not yet ready for prime time. :-(

**Comment by ilg-ul** [comment]
The pre-release is available from https://github.com/gnu-mcu-eclipse/arm-none-eabi-gcc/releases/tag/v8.2.1-1.3.

The final release to come soon.

**Comment by Trass3r** [comment]
At least historically ar never worked with LTO. You had to use gcc-ar which sets the right paths.

I've also struggled with the -static option killing the LTO plugin before. It's so weird that they coupled it like that, a bug imho.

So you say that bug report you opened also pertains to ARM's distribution? For my code it works.

**Comment by ilg-ul** [comment]
> You had to use gcc-ar which sets the right paths

ah, yes, I mean arm-none-eabi-ar (this is where everything started, @db4ple took the time to report). now the plugin is linked/copied to /lib/bfd-plugins, so this should be fixed.

> -static option ... so weird that they coupled it like that, a bug imho.

I agree. did you report it to GNU?

> pertains to ARM's distribution?

the bug pertains to the windows gcc-arm-none-eabi-8-2018-q4-major, and was created in the ARM tracker on launchpad.

it should occur if you create a project with the STM32F4 template, either C or C++. you can enable LTO from the template (last button). I'd not be surprised if projects for other STM families are also affected. 

for the moment I'm strugling to keep my Windows VM alive and tests on Windows are not a joy. :-(

**Comment by Trass3r** [comment]
> now the plugin is linked/copied to /lib/bfd-plugins, so this should be fixed.

Which seems to be the right thing to do:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=84995
https://gcc.gnu.org/viewcvs/gcc/trunk/gcc/doc/invoke.texi?r1=258953&r2=258952&pathrev=258953

> > -static option ... so weird that they coupled it like that, a bug imho.
>
> I agree. did you report it to GNU?

No. Please do, you know the process much better.

**Comment by ilg-ul** [comment]
I did further tests and the previous version (7-2018-q2-update) builds my projects properly with `-flto`.

So the problem was introduced in the latest release.

> No error with the PlatformIO STM32 Arduino package.

Are those C or C++?

**Comment by Trass3r** [comment]
Both. But if I add `-g` it fails with `3HvBdebugobjtem: file not recognized: file truncated`

**Comment by ilg-ul** [comment]
Ah, I already read about problems with debugging info.

I ran the tests again, and indeed, my projects fail only with -g3/-g, otherwise they build properly.

So, for now, LTO should not be used for Debug configurations.

**Comment by Trass3r** [comment]
Ok it all boils down to `arm-none-eabi-g++ -flto -g main.cpp` failing with an empty main.cpp!

**Comment by ilg-ul** [comment]
I thought that `arm-none-eabi-gcc -flto -g main.c` whould fail too, could you test it?

**Comment by Trass3r** [comment]
Yes fails too.

**Comment by ilg-ul** [comment]
Ok, so the conclusion stands, LTO should not be used for Debug configurations on Windows.

**Comment by ilg-ul** [comment]
> ... add to that bug ticket ...

good idea. done.

**Comment by Trass3r** [comment]
Thanks.
Interesting summary with known issues: http://hubicka.blogspot.com/2018/06/gcc-8-link-time-and-interprocedural.html

**Comment by ilg-ul** [comment]
indeed. after reading it, you understand why you can be happy if you make LTO work...

**Comment by Trass3r** [comment]
And why ARM's release schedule is way too slow.
gcc-9 will fix more LTO issues and also this major bug which prevents me from enabling -Werror but it will be 2020 until ARM releases that: https://bugs.launchpad.net/gcc-arm-embedded/+bug/1698539
Sigh they would just need to build gcc with `-DNO_IMPLICIT_EXTERN_C=1`...

**Comment by ilg-ul** [comment]
I just found a patch to fix the Windows LTO with -g/-g3, I'll test it right away and probably make a new release (1.4!)

so, if you have other improvements, please provide details and I'll try to accomodate them in 1.4.

**Comment by ilg-ul** [comment]
> -DNO_IMPLICIT_EXTERN_C=1

I took a look, but I'm a bit reluctant to add it to this release, since it may have unexpected effects, and I'd prefer Arm to test it thoroughly. I know that this might take quite a while...

**Comment by Trass3r** [comment]
Only that -isystem bug.

**Comment by Trass3r** [comment]
Yeah I know. Should you ever do a build with that change let me know and I'll test it :)
Which makes me wonder, did you try automatizing builds with something like AppVeyor?

**Comment by ilg-ul** [comment]
> did you try automatizing builds with something like AppVeyor?

I considered this, but I'm not sure it is realistic, multilib builds take several hours.

I have a super build script, that uses Docker, and builds all 4 binaries (linux/windows, 32/64) in one run. it takes more than one night to complete, but does everything. the project is public, you can take a look at https://github.com/gnu-mcu-eclipse/arm-none-eabi-gcc-build

**Comment by ilg-ul** [comment]
> Should you ever do a build with that change let me know and I'll test it :)

;-)

you can test it yourself, if you have a linux box at hand (I use an Ubuntu in a Virtual Box), install docker, clone the build scripts, tweak them and create your custom distribution.

**Comment by Trass3r** [comment]
I tried your repo on Windows but without a VM, i.e. using Docker for Windows directly or via WSL. But it's too tricky.

**Comment by ilg-ul** [comment]
> without a VM ... too tricky

I know, I did the same mistake on macOS. I don't know how well Docker on top of Windows works, but on top of macOS it has some major problems, the layer converting the macOS filesystem to the linux filesystem expected by Docker is very complicated, slow, and has memory leaks, for large builds it soon uses all available RAM and starts to use swap.

Knowing how weird Windows is and extrapolating, I would not be surprised if Docker has similar problems on Windows, or even worse.

So, the only reasonable environment for serious builds is a Linux, even a VM, it is faster and far more reliable to run Docker on it.

**Comment by Trass3r** [comment]
Well I checked again, crosstool-ng in WSL did build a toolchain, but there is no lto plugin, only a useless .a even after disabling static toolchain. And the linker complains about missing newlib_nano. Just saw you seem to copy them manually.

**Comment by ilg-ul** [comment]
Please contribute your findings to https://gcc.gnu.org/bugzilla/show_bug.cgi?id=89166, perhaps we can convince the maintainers to take a look at the bug, so far the initial response was not great.

**Comment by ilg-ul** [comment]
Fixed in 8.2.1-1.4 20190214.

**Comment by db4ple** [comment]
Thank you!!! Will test ASAP.
